### PR TITLE
WCiOS: Update Fastfile with version model

### PR DESCRIPTION
**Pull request by spencertransier**
## Summary
- This PR integrates the new version and build code models from https://github.com/wordpress-mobile/release-toolkit/pull/512
- It currently points to a development version of the Release Toolkit so that I could test out the call site implementation before releasing a new version of the Release Toolkit. 
- Nearly all of the `precheck` actions from the Release Toolkit were tightly coupled with the existing version methods. To simplify things with the new version model, I took the relevant snippets from the precheck actions and added them directly here. This is something that I've discussed with Olivier and Jeremy before in the context of "no one knows what the precheck actions do unless you dive deep down into the RT". These changes are somewhat opposite of DRY, but I think it's worth it for transparency. We may also decide in the future to create smaller actions that do more specific things versus the existing precheck and version bump actions that combine many disparate methods. 

## Before Merging
- [x] Point the Release Toolkit to a stable version

### CODE REVIEW

**Comment by spencertransier** [comment]
> Conflicting files
>
> Gemfile.lock

I will get this sorted out when there is a new Release Toolkit version to update to

File: fastlane/Fastfile
Comment (spencertransier) [inline fastlane/Fastfile:1294]: I created a test lane that verified these values are being calculated and formatted as expected (given a `VERSION_LONG` of `15.5.2` and a `VERSION_SHORT` of `15.5.2.4`):

Screenshot 2023-09-28 at 11 35 30 AM

File: fastlane/Fastfile
Comment (spencertransier) [inline fastlane/Fastfile:425]: After searching the release scenario, Buildkite scripts, and Buildkite pipelines, I did not find any instances of this lane being used.

File: fastlane/Fastfile
Comment (spencertransier) [inline fastlane/Fastfile:452]: After searching the release scenario, Buildkite scripts, and Buildkite pipelines, I did not find any instances of this lane being used.

File: fastlane/Fastfile
Comment (spencertransier) [inline fastlane/Fastfile:41]: This is where the various version and build code formatters/calculators are instantiated to be used by the rest of the Fastfile.

File: fastlane/Fastfile
Comment (spencertransier) [inline fastlane/Fastfile]: I moved this check to the top of the `code_freeze` lane instead of `before_all`. It hasn't been useful to have it run before every single lane during the course of a release.

File: fastlane/Fastfile
Comment (AliSoftware) [inline fastlane/Fastfile]: ~~I'm not sure I understand what this is supposed to do?~~

~~When  I look at the implementation for this method from what i can see it only does a check then adds a sentence to the `message`… but doesn't actually _git switch_ to a specific branch… So either the naming of that method is misleading, or it doesn't do what it's supposed to?~~

[EDIT] My bad, I now see that it in fact does `checkout_and_pull` after all, missed it in my initial read of that method's implementation. See https://github.com/woocommerce/woocommerce-ios/pull/10807#discussion_r1341722951 below.

File: fastlane/Fastfile
Comment (AliSoftware) [inline fastlane/Fastfile]: Fastlane has an action called `ensure_git_branch` that is built in, we should use this instead.

Also, one thing to know about is that our toolkit's `ensure_on_branch!` checks the branch name using `include?`; while the fastlane one (`ensure_git_branch`) does the check using `RegExp.match?`

This means that when you use `ensure_git_branch` you can even use `ensure_git_branch(branch: '^release/')` to check if the branch _starts_ with `release/`, which will match `release/12.3` but not `merge/release-into-trunk` for example.

File: fastlane/Fastfile
Comment (AliSoftware) [inline fastlane/Fastfile]: I can't see a method named `build_code` defined anywhere? Was this supposed to be `BUILD_CODE_FORMATTER.build_code(version: new_version)` instead?

File: fastlane/Fastfile
Comment (AliSoftware) [inline fastlane/Fastfile:617]: Missing a value for this parameter here 🤔

File: fastlane/Fastfile
Comment (AliSoftware) [inline fastlane/Fastfile]: Oh ok my bad, it does call `checkout_and_pull`, it's just that this is kinda hidden within the `unless` clause (I didn't expect any "action with side effect" to be in the `unless` clause, as opposed to just a boolean value, that's why I missed it)

Maybe to help readability we should rewrite this like that?
```suggestion
def switch_to_user_branch(version, message)
  success = Fastlane::Helper::GitHelper.checkout_and_pull(release: version)
  UI.user_error!("Release branch for version #{version} doesn't exist. Abort.") unless success
  message << "Looking at branch release/#{version} as requested by user. Detected version: #{current_build_code}.\n"
end
```

Also, I'm not a fan of having a method that modifies one of its input parameters (`message`), as opposed to return it as a new value. Doesn't feel very idiomatic of Ruby.

File: fastlane/Fastfile
Comment (spencertransier) [reply fastlane/Fastfile]: ~Good to know! I'm pretty sure I spotted some uses of the built-in action. I'm going to hold off on changing all of these over to the built-in action for a separate PR. I've tried to just replicate everything that was in the precheck and version bump actions so that we reduce the possibility of introducing unexpected behavior with the rest of these changes.~

File: fastlane/Fastfile
Comment (spencertransier) [reply fastlane/Fastfile]: That was left over from an earlier implementation of the build_code methods. I'll update it to `next_build_code` instead.

File: fastlane/Fastfile
Comment (spencertransier) [reply fastlane/Fastfile:617]: That's the Ruby 3 convention of not needing to include the argument if the name of the argument is the same as the name of the parameter.

File: fastlane/Fastfile
Comment (spencertransier) [reply fastlane/Fastfile]: Addressed in https://github.com/woocommerce/woocommerce-ios/commit/be0e00af7dcddbf962ccca507f0097815a128c46

File: fastlane/Fastfile
Comment (spencertransier) [reply fastlane/Fastfile]: Addressed in https://github.com/woocommerce/woocommerce-ios/commit/be0e00af7dcddbf962ccca507f0097815a128c46

File: fastlane/Fastfile
Comment (spencertransier) [reply fastlane/Fastfile]: Addressed in https://github.com/woocommerce/woocommerce-ios/commit/be0e00af7dcddbf962ccca507f0097815a128c46

File: fastlane/Fastfile
Comment (spencertransier) [reply fastlane/Fastfile]: Addressed here: https://github.com/woocommerce/woocommerce-ios/pull/10807/commits/b8802015356361954a34b2f3ea6e0241483966a7

**Review by mokagio — COMMENTED** [review]
Just a note that I've been focusing on https://github.com/wordpress-mobile/WordPress-iOS/pull/21666 and leaving various suggestions there. 

I think it'd be best to lock in the implementation on that end, then propagate the learnings here.

**Comment by spencertransier** [comment]
@mokagio I've updated this one with the changes from WP/JPiOS, so it is ready to review again.

File: fastlane/Fastfile
Comment (mokagio) [inline fastlane/Fastfile]:
```suggestion
    UI.user_error!('Aborted by user request') unless options[:skip_confirm] || UI.confirm('Do you want to continue?')
```

File: fastlane/Fastfile
Comment (mokagio) [inline fastlane/Fastfile]: How do you feel about removing this check? What are the chances the branch changed under the hood after the `checkout_and_pull` command a few lines above?

```suggestion
```

File: fastlane/Fastfile
Comment (mokagio) [inline fastlane/Fastfile]: I run into an issue with this during the 23.6 Jetpack/WordPress code freeze and ended up replacing it with `git_commit` in https://github.com/wordpress-mobile/WordPress-iOS/pull/21911/commits/f886436c4fcdbaf6d4fd3a1f65e06a3fc06b0b8e

More info in https://github.com/wordpress-mobile/release-toolkit/issues/525

How would you feel about doing the same here? Otherwise, because of the `ENV['PROJECT_ROOT_FOLDER']` above, I fear this will break on the next code freeze.

File: fastlane/Fastfile
Comment (mokagio) [inline fastlane/Fastfile]:
```suggestion
    UI.user_error!('Aborted by user request') unless options[:skip_confirm] || UI.confirm('Do you want to continue?')
```

File: fastlane/Fastfile
Comment (mokagio) [inline fastlane/Fastfile]:
```suggestion
    UI.user_error!('Aborted by user request') unless options[:skip_confirm] || UI.confirm('Do you want to continue?')
```

File: fastlane/Fastfile
Comment (mokagio) [inline fastlane/Fastfile]: In the interest of failing early, should we move this at the start of the lane?

File: fastlane/Fastfile
Comment (mokagio) [inline fastlane/Fastfile]: I don't remember where we discussed this already, but I if I recall correctly the conclusion was that the push command should have remained in the lane?

Otherwise, we might run into a scenario where the release manager has changes locally that they assume would be pushed (like the other finalization lanes do) but the CI build is triggered on the latest commit on the remote.

Or... did we conclude that there should be no local change because all changes in hotfixes happen via PRs and are already in the remote?

File: fastlane/Fastfile
Comment (spencertransier) [reply fastlane/Fastfile]: Fixed in https://github.com/woocommerce/woocommerce-ios/commit/01da9abb2de7063634358f8f19c9f89a0009376a

File: fastlane/Fastfile
Comment (spencertransier) [reply fastlane/Fastfile]: That's a great point. I don't think I was a part of that discussion, but I think that scenario makes sense. _Most likely_, all the changes will be done as PRs, but might as well push any local changes just in case. 

I added the push command in https://github.com/woocommerce/woocommerce-ios/commit/8a70f61035d4d9de7c120791bbfb210e1d716f62

File: fastlane/Fastfile
Comment (spencertransier) [reply fastlane/Fastfile]: Shoot, I missed that method during the env var project because it referenced the env var like `get_from_env!(key: 'PROJECT_ROOT_FOLDER')` instead of `ENV['PROJECT_ROOT_FOLDER']`. Sorry, I should have caught that. I'll make sure to get the other iOS apps updated. 

Fixed for this PR in https://github.com/woocommerce/woocommerce-ios/commit/97204c6a0430e90c290e9d72147eb9d993b8af4d

File: fastlane/Fastfile
Comment (spencertransier) [reply fastlane/Fastfile]: Fixed in https://github.com/woocommerce/woocommerce-ios/pull/10807/commits/5cd92da003c3a407bcd0c675b63c72db14c8c87e

**Comment by spencertransier** [comment]
@mokagio Thank you for taking another look! Sorry, I thought I had already fixed most of those cases for WCiOS, but I think I was getting it mixed up with another app 🤪. All your comments are now addressed 👍

**Review by mokagio — APPROVED** [review]
All looks good. At this point, the best thing to do is to merge and see how we go on the next code freeze 👍

**Pull request merged**
